orca-bakery/src/main/groovy/com/netflix/spinnaker/orca/bakery/pipeline/BakeStage.groovy

```diff
@@ -129,7 +136,28 @@ class BakeStage implements StageDefinitionBuilder {
           return deploymentDetails
         }
       ]
+
+      // TODO: add condition for user-specified opt-out setting in addition to feature toggle
+      if (failOnImageNameMismatch()) {
+        def imageNames = globalContext.deploymentDetails.stream().map({details -> details['imageName']})
+          .distinct()
+          .collect(Collectors.toList())
+
+        if (imageNames.size() > 1) {
+          throw new IllegalStateException("Non-matching image names found in different regions: ${imageNames}. " +
+          "Failing bake stage as this would cause deployment failures. Please re-run the bake.")
+        }
+      }
+
       TaskResult.builder(ExecutionStatus.SUCCEEDED).outputs(globalContext).build()
     }
+
+    def failOnImageNameMismatch() {
```

Comment:
Avoid using `def` at all times in Groovy's `main` source set, especially in method contracts.

Reply:
Any particular reason why?

Reply:
ideally, we want to be on Java/Kotlin instead of groovy. When we make changes to the groovy code there are two options: 
1. convert to Java
2. keep it in groovy but make it as java as possible (all valid java code is valid groovy code)

and... type inference like this is hard to read and can be error prone

Reply:
(oh, and as Rob said, this applies to `main` not `test` source sets

Reply:
FYI backend conventions are here: https://www.spinnaker.io/community/contributing/back-end-code/

Reply:
Thanks for the context @marchello2000. In that case, shouldn't we apply that rule to any declaration (not just methods)? I see a bunch of `def variable` all over, and was declaring my variables in the same way. Should I always declare them with a type then?

Reply:
Ideally, yes, I try not to use `def`s at all in any new/changed groovy code (other than tests)